File: SwiftySheetsDemo/Parser/AST/ASTStyleDefinition.m
//
//  ASTStyleDefinition.m
//  SwiftySheetsDemo
//
//  Created by imihaly on 09/07/16.
//  Copyright © 2016 Agnes Vasarhelyi. All rights reserved.
//

#import "ASTStyleDefinition.h"

@interface ASTStyleDefinition()

@property(nonatomic, copy, readwrite) NSString *name;
@property(nonatomic, strong, readwrite) NSMutableArray *children;
@property(nonatomic, strong, readwrite) NSMutableArray *properties;

@end

@implementation ASTStyleDefinition

+ (instancetype)definitionWithName:(NSString *)name elements:(NSArray *)elements {
	return [[self alloc] initWithName:name elements:elements];
}

- (instancetype)initWithName:(NSString *)name elements:(NSArray *)elements {
	self = [super init];
	if(self) {
		self.name = name;
		self.children = NSMutableArray.array;
		self.properties = NSMutableArray.array;
		for(id element in elements) {
			if([element isKindOfClass:ASTPropertyDefinition.class]) {
				[self addPropertyDefinition:(ASTPropertyDefinition *)element];
			} else if([element isKindOfClass:ASTStyleDefinition.class]) {
				[self addChildDefinition:(ASTStyleDefinition *)element];
			}
		}
	}
	return self;
}

- (void)addPropertyDefinition:(ASTPropertyDefinition *)propertyDefinition {
	[self.properties addObject:propertyDefinition];
}

- (void)addChildDefinition:(ASTStyleDefinition *)styleDefinition {
	[self.children addObject:styleDefinition];
}

- (ASTStyleDefinition *)styleDefinitionWithName:(NSString *)name {
	for(ASTStyleDefinition *def in self.children) {
		if([def.name isEqualToString:name]) return def;
	}
	return nil;
}

- (ASTPropertyDefinition *)propertyDefinitionWithName:(NSString *)name {
	for(ASTPropertyDefinition *def in self.properties) {
		if([def.propertyName isEqualToString:name]) return def;
	}
	return nil;
}


- (NSString *)description {
	return [NSString stringWithFormat:@"%@{\n%@ \n%@\n}" ,self.name, self.properties, self.children];
}

@end

File: SwiftySheetsDemo/Parser/AST/ASTStyleDefinition.swift
//
//  ASTStyleDefinition.swift
//  SwiftySheetsDemo
//
//  Created by Alföldi Norbert on 19/07/16.
//  Copyright © 2016 Agnes Vasarhelyi. All rights reserved.
//

import Foundation

public class ASTStyleDefinition : NSObject {
	public var name: String
	public var children = Array<ASTStyleDefinition>()
	public var properties = Array<ASTPropertyDefinition>()
	
	public init(name: String, elements: Array<AnyObject>) {
		self.name = name
		
		for element in elements where element is ASTStyleDefinition {
			self.children.append(element as! ASTStyleDefinition)
		}
		
		for element in elements where element is ASTPropertyDefinition {
			self.properties.append(element as! ASTPropertyDefinition)
		}
	}
	
	public func addStyleDefinition(styleDefinition: ASTStyleDefinition) {
		self.children.append(styleDefinition)
	}
	
	public func addPropertyDefinition(propertyDefinition: ASTPropertyDefinition) {
		self.properties.append(propertyDefinition)
	}
	
	public func styleDefinitionWithName(name: String) -> ASTStyleDefinition? {
		return self.children.filter({ $0.name == name }).first
	}
	
	public func propertyDefinitionWithName(name: String) -> ASTPropertyDefinition? {
		return self.properties.filter({ $0.propertyName == name }).first
	}
}
